File: linda-core/src/main/java/it/unibo/coordination/utils/events/SyncOrderedEventSourceImpl.java
package it.unibo.coordination.utils.events;

import java.util.Collection;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

class SyncOrderedEventSourceImpl<T> extends AbstractEventSourceImpl<T> {

    private final List<EventListener<T>> eventListeners = new LinkedList<>();

    @Override
    protected Collection<EventListener<T>> getEventListeners() {
        return eventListeners;
    }

    @Override
    public T syncEmit(T data) {
        for (var listener : eventListeners) {
            listener.onEvent(data);
        }
        return data;
    }

    @Override
    public CompletableFuture<T> asyncEmit(T event) {
        return CompletableFuture.completedFuture(syncEmit(event));
    }


}

File: utils/src/main/kotlin/it/unibo/coordination/utils/events/SyncOrderedEventSourceImpl.kt
package it.unibo.coordination.utils.events

import java.util.*
import java.util.concurrent.CompletableFuture

internal class SyncOrderedEventSourceImpl<T> : AbstractEventSourceImpl<T>() {

    override val eventListeners = LinkedList<EventListener<T>>()

    override fun syncEmit(data: T): T {
        for (listener in eventListeners) {
            listener(data)
        }
        return data
    }

    override fun asyncEmit(event: T): CompletableFuture<T> {
        return CompletableFuture.completedFuture(syncEmit(event))
    }

}